vectorizers/preprocessing.py

```diff
@@ -666,3 +692,373 @@ def preprocess_token_sequences(
+            max_unique_tokens=max_unique_tokens,
+            min_frequency=min_frequency,
+            max_frequency=max_frequency,
+            min_occurrences=min_occurrences,
+            max_occurrences=max_occurrences,
+            min_document_frequency=min_document_frequency,
+            max_document_frequency=max_document_frequency,
+            min_document_occurrences=min_document_occurrences,
+            max_document_occurrences=max_document_occurrences,
+            total_tokens=total_tokens,
+            total_documents=len(token_sequences),
+        )
+
+    if masking is None:
+        result_sequences = List()
+        for i, sequence in enumerate(token_sequences):
+            result_sequences.append(
+                np.array(
+                    [
+                        (token_dictionary[token[0]], token[1])
+                        for token in sequence
+                        if token[0] in token_dictionary
+                    ],
+                    dtype=np.float32,
+                )
+            )
+    else:
+        result_sequences = List()
+        if masking in token_dictionary:
+            del token_dictionary[masking]
+
+        for i, sequence in enumerate(token_sequences):
+            result_sequences.append(
+                np.array(
+                    [
+                        (len(token_dictionary), token[1])
+                        if not (token[0] in token_dictionary)
+                        else (token_dictionary[token], token[1])
+                        for token in sequence
+                    ],
+                    dtype=np.float32,
+                )
+            )
+        token_dictionary[masking] = len(token_dictionary)
+
+    inverse_token_dictionary = {
+        index: token for token, index in token_dictionary.items()
+    }
+
+    return (
+        result_sequences,
+        token_dictionary,
+        inverse_token_dictionary,
+        token_frequencies,
+    )
+
+
+def preprocess_multi_token_sequences(
+    token_sequences,
+    token_dictionary=None,
+    max_unique_tokens=None,
+    min_occurrences=None,
+    max_occurrences=None,
+    min_frequency=None,
+    max_frequency=None,
+    min_document_occurrences=None,
+    max_document_occurrences=None,
+    min_document_frequency=None,
+    max_document_frequency=None,
+    ignored_tokens=None,
+    excluded_token_regex=None,
+    masking=None,
+):
+    """Perform a standard set of preprocessing for token sequences. This includes
+    constructing a token dictionary and token frequencies, pruning the dictionary
+    according to frequency and ignored token constraints, and editing the token
+    sequences to only include tokens in the pruned dictionary. Note that either
+    min_occurrences or min_frequency can be provided (respectively
+    max_occurences or max_frequency). If both are provided they must agree.
+
+    Parameters
+    ----------
+    token_sequences: Iterable of (tuple | list | numpy.array)
```

Comment:
Should this be a "a list of multitoken sequences" and a few more details about the data type?